dm/syncer/dml_worker.go

```diff
@@ -117,24 +115,24 @@ func (w *DMLWorker) run() {
 
 	for j := range w.inCh {
 		metrics.QueueSizeGauge.WithLabelValues(w.task, "dml_worker_input", w.source).Set(float64(len(w.inCh)))
-		if j.tp == flush || j.tp == conflict {
-			if j.tp == conflict {
-				w.addCountFunc(false, adminQueueName, j.tp, 1, j.targetTable)
-			}
-			w.wg.Add(w.workerCount)
-			// flush for every DML queue
-			for i, jobCh := range jobChs {
-				startTime := time.Now()
-				jobCh <- j
-				metrics.AddJobDurationHistogram.WithLabelValues(j.tp.String(), w.task, queueBucketMapping[i], w.source).Observe(time.Since(startTime).Seconds())
-			}
-			w.wg.Wait()
-			if j.tp == conflict {
-				w.addCountFunc(true, adminQueueName, j.tp, 1, j.targetTable)
-			} else {
-				w.flushCh <- j
-			}
-		} else {
+		switch j.tp {
+		case flush:
+			w.addCountFunc(false, adminQueueName, j.tp, 1, j.targetTable)
+			w.sendJobToAllDmlQueue(j, jobChs, queueBucketMapping)
+			j.wg.Wait()
+			w.addCountFunc(true, adminQueueName, j.tp, 1, j.targetTable)
+			w.flushCh <- j
+		case asyncFlush:
+			w.addCountFunc(false, adminQueueName, j.tp, 1, j.targetTable)
+			w.sendJobToAllDmlQueue(j, jobChs, queueBucketMapping)
+			w.addCountFunc(true, adminQueueName, j.tp, 1, j.targetTable)
```

Comment:
because async doesn't need `j.wg.Wait()`, this is very quickly and should almost be same with `metrics.AddJobDurationHistogram`.

Also for DML jobs, the `w.addCountFunc(true, ...)` is postponed to Syncer.successFunc, so for async flush it's more consistent that we also `w.addCountFunc(true, ...)` when async flush is finished?